File: Baekjoon/17299 - 오등큰수/17299.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <map>
using namespace std;

#define FASTIO cin.tie(NULL); cout.tie(NULL); ios::sync_with_stdio(false);
#define MAX 1111111

int main() {
    FASTIO

    int n, temp;
    vector<int> vec, result(MAX, -1), cnt(MAX, 0);
    stack<int> st;

    cin >> n;

    for(int i = 0; i < n; i++) {
        cin >> temp;

        vec.push_back(temp);
        cnt[temp] += 1;
    }

    for(int i = 0; i < n; i++) {
        while(st.size() && cnt[vec[st.top()]] < cnt[vec[i]]) {
            result[st.top()] = vec[i];
            st.pop();
        }
        st.push(i);
    }

    for(int i = 0; i < n; i++) {
        cout << result[i] << " ";
    }
}

File: solution/17xxx/17299 - 오등큰수/17299.py
import sys
input = sys.stdin.readline

MAX = 1111111

result = [-1]*MAX
cnt = [0]*MAX
st = []

n = int(input())
a = list(map(int, input().split()))

for i in range(n):
    cnt[a[i]] += 1

for i in range(n):
    while len(st) != 0 and cnt[a[st[-1]]] < cnt[a[i]]:
        result[st[-1]] = a[i]
        st.pop()
    st.append(i)

for i in range(n):
    print(result[i], end=" ")